### PR TITLE
Add custom index information

### LINKED ISSUE #207: Suggested additional indexes
#16 suggests non-standard indexes.

What non-standard indexes should we be suggesting for performance reasons?

```sql
CREATE INDEX water_areas_idx ON planet_osm_polygon USING gist (way) 
  WHERE (((waterway IS NOT NULL) 
      OR (landuse = ANY (ARRAY['reservoir'::text, 'water'::text, 'basin'::text]))) 
    OR ("natural" IS NOT NULL));
```

```sql
CREATE INDEX ferry_idx ON planet_osm_line USING gist (way) WHERE (route = 'ferry'::text)
```

### PR DESCRIPTION
**Pull request by pnorman**
These custom indexes should speed up rendering, particularly at low and middle zooms.

Fixes #207

cc @woodpeck @jburgess777 @Firefishy 

I'm not recommending any way_area indexes, not because I don't think they're worth it, but because they're more complicated to get right, and I'd rather go ahead and recommend a number of indexes I already know help performance.

### CODE REVIEW

**Comment by pnorman** [comment]
Suggested to me was an index for low zoom admin polygons to reflect the switch to OSM data for low zoom labeling. My benchmarking was prior to this, but I recall it being a good idea.

**Comment by rrzefox** [comment]
The `IF NOT EXISTS` will only work for postgresql >= 9.5. 9.4 or older do not have it (so in particular, the 9.3 in ubuntu 14.04 does not have it).

**Comment by pnorman** [comment]
> The IF NOT EXISTS will only work for postgresql >= 9.5. 9.4 or older do not have it (so in particular, the 9.3 in ubuntu 14.04 does not have it).

Updated.

File: indexes.sql [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  ON planet_osm_point USING gist (way)
+  WHERE place IN ('city', 'town') AND name IS NOT NULL;
+
+CREATE INDEX planet_osm_point_place
+  ON planet_osm_point USING gist (way)
+  WHERE place IS NOT NULL AND name IS NOT NULL;
+
+CREATE INDEX planet_osm_point_name
+  ON planet_osm_point USING gist (way)
+  WHERE name IS NOT NULL;
```
Comment (pnorman) [inline indexes.sql]: This index is 653MB and only used by `text-point`, so I'm not this is worth the additional memory cache pressure.

**Comment by pnorman** [comment]
I have a WIP changing how these are defined to allow creation of SQL statements to reindex, etc.

**Comment by pnorman** [comment]
* `indexes.sql` is now a generated file, but shouldn't need to be changed much
* The script options are
```
  -h, --help            show this help message and exit
  --concurrent          Generate indexes CONCURRENTLY
  --fillfactor FILLFACTOR
                        Custom fillfactor to use
  --notexist            Use IF NOT EXISTS (requires 9.5)
  --osm2pgsql           Include indexes normally built by osm2pgsql
  --reindex             Rebuild existing indexes
```

Examples are
default:
```sql
CREATE INDEX planet_osm_roads_admin
  ON planet_osm_roads USING GIST (way)
  WHERE boundary = 'administrative';
```
``--fillfactor 100``:
```sql
CREATE INDEX planet_osm_roads_admin
  ON planet_osm_roads USING GIST (way)
  WITH (fillfactor=100)
  WHERE boundary = 'administrative';
```
``--concurrent``:
```sql
CREATE INDEX CONCURRENTLY planet_osm_roads_admin
  ON planet_osm_roads USING GIST (way)
  WHERE boundary = 'administrative';
```
``--reindex``:
```sql
REINDEX planet_osm_roads_admin;
```
``--reindex --concujrrent``
```sql
ALTER INDEX planet_osm_roads_admin RENAME TO planet_osm_roads_admin_old;
CREATE INDEX CONCURRENTLY planet_osm_roads_admin
  ON planet_osm_roads USING GIST (way)
  WHERE boundary = 'administrative';
DROP INDEX planet_osm_roads_admin_old;
```

`--osm2pgsql` simply adds statements for `planet_osm_[point|line|polygon|roads]_index`

File: indexes.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    where: landuse = 'military'
+roads:
+  # The roads table only has a subset of data, so it's just got some low-zoom
+  # indexes and some fairly selective ones for high zoom
+  admin_low:
+    where: boundary = 'administrative' AND admin_level IN ('0', '1', '2', '3', '4')
+  admin:
+    where: boundary = 'administrative'
+  roads_ref:
+    where: WHERE highway IS NOT NULL AND ref IS NOT NULL
```
Comment (pnorman) [inline indexes.yml]: shouldn't have the WHERE here

File: indexes.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# A goal with the indexes is to have them general-purpose enough to not need
+# frequent changing with stylesheet changes, but to be usable with many
+# versions, and potentially other styles.
+
+point:
+  # The point table is small, but so are the partial indexes
+  place_low:
+    where: place IN ('city', 'town') AND name IS NOT NULL
```
Comment (pnorman) [inline indexes.yml]: turns out this isn't worth it

**Comment by pnorman** [comment]
I think this is ready

**Comment by math1985** [comment]
Thanks!

**Pull request merged**